pkg/dependency/parser/golang/binary/parse.go

```diff
@@ -98,14 +85,47 @@ func (p *Parser) Parse(r xio.ReadSeekerAt) ([]ftypes.Package, []ftypes.Dependenc
 			mod = dep.Replace
 		}
 
+		version := p.checkVersion(mod.Path, mod.Version)
 		pkgs = append(pkgs, ftypes.Package{
-			Name:    mod.Path,
-			Version: p.checkVersion(mod.Path, mod.Version),
+			ID:           dependency.ID(ftypes.GoBinary, mod.Path, version),
+			Name:         mod.Path,
+			Version:      version,
+			Relationship: ftypes.RelationshipUnknown,
 		})
 	}
-
 	sort.Sort(pkgs)
-	return pkgs, nil, nil
+
+	// There are times when gobinaries don't contain Main information.
+	// e.g. `Go` binaries (e.g. `go`, `gofmt`, etc.)
+	var deps []ftypes.Dependency
+	if info.Main.Path != "" {
+		// Only binaries installed with `go install` contain semver version of the main module.
+		// Other binaries use the `(devel)` version, but still may contain a stamped version
+		// set via `go build -ldflags='-X main.version=<semver>'`, so we fallback to this as.
+		// as a secondary source.
+		// See https://github.com/aquasecurity/trivy/issues/1837#issuecomment-1832523477.
+		version := cmp.Or(p.checkVersion(info.Main.Path, info.Main.Version), p.ParseLDFlags(info.Main.Path, ldflags))
+		root := ftypes.Package{
+			ID:           dependency.ID(ftypes.GoBinary, info.Main.Path, version),
+			Name:         info.Main.Path,
+			Version:      version,
+			Relationship: ftypes.RelationshipRoot,
+		}
+		deps = []ftypes.Dependency{
+			{
+				ID: root.ID,
+				// Consider all packages as dependencies of the main module.
+				DependsOn: lo.Map(pkgs, func(pkg ftypes.Package, _ int) string {
+					return pkg.ID
+				}),
+			},
+		}
+		// Add main module
+		pkgs = append(pkgs, root)
+		sort.Sort(pkgs)
```

Comment:
We sort `pkgs` twice.
Can we sort them at the end of the function?

```Go
sort.Sort(pkgs)
return pkgs, deps, nil
```

Reply:
But most likely in this case we need to sort `deps`.

Reply:
We need to sort `dependsOn` anyway. It requires additional three lines because `sort.Strings` doesn't return `[]string`, but sorting packages requires one line. Do you prefer sorting `dependsOn`?

Reply:
`Package` sort checks multiple fields (relations, name, etc.) and moves structures inside slice.
on the other hand `sort.Strings` works only with 1 string for each slice element.
but I'm not sure if it will affect performance much.

Another point - usually we use `sort.Strings` for `dependsOn`

But overall i don't mind your approach

Reply:
OK, updated.
https://github.com/aquasecurity/trivy/pull/7973/commits/2a5e77b632a46843cf249a44b19a5a2f8fe5d0d8

Reply:
hm... your solution has one advantage - `stdlib` is first element in the slice (because `stdlib` is root, other dependencies are `unknown`)

```
expected: []types.Dependency{types.Dependency{ID:"github.com/ebati/trivy-mod-parse", DependsOn:[]string{"stdlib@v1.16.4", "github.com/davecgh/go-spew@v1.1.1", "github.com/go-sql-driver/mysql@v1.5.0"}}}
        	            	actual  : []types.Dependency{types.Dependency{ID:"github.com/ebati/trivy-mod-parse", DependsOn:[]string{"github.com/davecgh/go-spew@v1.1.1", "github.com/go-sql-driver/mysql@v1.5.0", "stdlib@v1.16.4"}}}
```